manager/volume.go

```diff
@@ -614,7 +617,58 @@ func (m *VolumeManager) Expand(volumeName string, size int64) (v *longhorn.Volum
 	}
 
 	if v.Spec.Size > size {
-		return nil, fmt.Errorf("cannot expand volume %v with current size %v to a smaller size %v", v.Name, v.Spec.Size, size)
+		return nil, fmt.Errorf("longhorn volume %v spec size %v is greater than requested Persistent Volume Claim size %v", v.Name, v.Spec.Size, size)
+	}
```

Comment:
If we want to keep this behavior, you should move it below the PVC processing.
Since that way the user can use the longhorn api to update PVC, even if the volume is already bigger.

Reply:
I personally feel we should return success in this case instead of an error, since the API is called Expand instead of Resize.
It's true we can never shrink a volume, but I also don't see why we couldn't return success as in:

- We ensure that the capacity of the volume is at least the requested user capacity.
- In a sense we already operate on in this way, since we need to round the user supplied capacity up to a multiple of our block size.
- The other benefit of this change is that we are in sync with kubernetes behavior and don't need the special casing in the csi driver.

We discussed this previously though and since it's a change in behavior we agreed on not changing it at that point in time.
Now that we are adding additional behavior to this API I think it's fair to also consider this case. Since the api behavior is changing already -> i.e early exit because of PVC, relying on the CSI driver to update the longhorn volume size.

Reply:
cc @shuo-wu @innobead

Reply:
My main concern is to make sure that when user call API, workload/volume will not go into a state where it cannot attach as in the issue.
If I remember correctly, when PVC Size < LHV Size, workload's volume attachment will fail.  Therefore, request size need to be greater than LHV Size.
If we move this check after PVC expand and request size is greater than PVC size and less than LHV size, the scenario will still occur.  
I know it is good by design to move it after PVC call, but do we want to leave it to the user to resolve the issue manually?

Reply:
I don't understand your concern, can you clarify.

These are the 2 cases that require updates:
- if pvc.size < requestedSize: -> update PVC
- if volume.size < requestedSize: -> update volume

From the logic these 2 should be handled in this order:
- PVC evaluation & update
- Volume evaluation & update

You can even do both of the updates in the same api call, since the expansion call needs to be idempotent.

Reply:
I know it might not be a normal UI case, but I am a little hackish and I like to edit manifest through scripting and use UI as last resort. (And yes, UI has done some prevention) and  I am not sure if there are other scenarios like this.

Assuming original LHV/PVC size is 1Gi, editing Longhorn volume by manifest

```
kubectl patch -p '{"spec":{"size": "3Gi"}}' volume.longhorn.io ${LHV} --namespace longhorn-system --type=merge
```

See that PV/PVC didn't update and try to call longhorn API

```
curl -sSX POST -H 'Content-Type: application/json' "http://${BIND_URL}:${BIND_PORT}/v1/volumes/${PV}?action=${ACTION}" -d '{"size":"2Gi"}' | jq
```

In such scenario, request size will be greater than PVC size but less than LHV size. With current implementation, the user will receive an error.  If move after PVC checking, I remember PVC request will change but PVC/PV status will never update nor receive error. UI will display LHV as 3Gi, but workload would not able to attached.  Need to increase request size to 4Gi in UI for workload/volume to get back to normal state.

Reply:
So what happens if you do the above? I don't expect any problem, if there is then that is a bug.
- Create Volume (1G)
- Workload scale down
- Edit lh.spec.Size via kubectl (3G)
- Wait for expansion, you can scale up before, but it will not be ready
- Workload scale up (should work, even though LH size > pvc size)
- Workload scale down
- Edit pvc.size via kubectl (2G)
- this should go through the csi expansion process
- return success PV should become (3G) while pvc (2G)
- Workload scale up (should work, even though LH size > pvc size)

Are you seeing any failure in the above process?

Reply:
Considering the scenario Joshua mentioned, allowing Longhorn to return success when the requested size is smaller than the volume spec size makes sense.

Reply:
Update the code as (in the latest code)

> if pvc.size < requestedSize: -> update PVC
> if volume.size < requestedSize: -> update volume

And go through the steps

> Create Volume (1Gi)
> Workload scale down
> Edit lh.spec.Size via kubectl (3Gi)
> Wait for expansion, you can scale up before, but it will not be ready
> Workload scale up (should work, even though LH size > pvc size)
> Workload scale down
> Edit pvc.size via kubectl (2Gi)
> this should go through the csi expansion process
> return success PV should become (3Gi) while pvc (2Gi)
> Workload scale up (should work, even though LH size > pvc size)

Workload is able to attached volume and see storage volume as following in the end:

```
PVC
resources.request.storage = 2Gi
status.capacity.storage = 3Gi
PV
status.capacity.storage = 3Gi
LHV
spec.size=3221225472
```

Reply:
Great as shuo mentioned above we can combine the v.size > requested size and v.size == requestedSize to return the current volume instead of error.

Let's have a quick sync up tomorrow morning my time then we can merge this. @shuo-wu anything missing from your perspective?

Reply:
updated